src/app/loadout-builder/mod-utils.ts

```diff
@@ -31,3 +41,299 @@ export function someModHasEnergyRequirement(mods: PluggableInventoryItemDefiniti
+  const energyUsedAndWasted = modCost + itemEnergy.originalCapacity;
+  const energyInvested = Math.max(0, modCost - itemEnergy.originalCapacity);
+
+  return finalEnergy === itemEnergy.originalType ? energyInvested : energyUsedAndWasted;
+}
+
+function getItemEnergyType(
+  defs: D2ManifestDefinitions,
+  item: DimItem,
+  upgradeSpendTier: UpgradeSpendTier,
+  lockItemEnergyType: boolean,
+  bucketSpecificMods?: PluggableInventoryItemDefinition[]
+) {
+  if (!item.energy) {
+    return DestinyEnergyType.Any;
+  }
+
+  const bucketSpecificModType = bucketSpecificMods?.find(
+    (mod) => mod.plug.energyCost && mod.plug.energyCost.energyType !== DestinyEnergyType.Any
+  )?.plug.energyCost?.energyType;
+
+  // if we find bucket specific mods with an energy type we have to use that
+  if (bucketSpecificModType) {
+    return bucketSpecificModType;
+  }
+
+  return canSwapEnergyFromUpgradeSpendTier(defs, upgradeSpendTier, item, lockItemEnergyType)
+    ? DestinyEnergyType.Any
+    : item.energy.energyType;
+}
+
+function energyTypesAreCompatible(first: DestinyEnergyType, second: DestinyEnergyType) {
+  return first === second || first === DestinyEnergyType.Any || second === DestinyEnergyType.Any;
+}
+
+export function getModAssignments(
+  items: DimItem[],
+  mods: PluggableInventoryItemDefinition[],
+  defs: D2ManifestDefinitions | undefined,
+  upgradeSpendTier: UpgradeSpendTier,
+  lockItemEnergyType: boolean
+): Map<string, PluggableInventoryItemDefinition[]> {
+  if (!defs) {
+    return new Map();
+  }
+
+  const bucketSpecificAssignments = new Map<string, PluggableInventoryItemDefinition[]>();
+  const bucketIndependantAssignments = new Map<string, PluggableInventoryItemDefinition[]>();
+  // just an arbitrarily large number
+  let assignmentEnergyCost = 10000;
+
+  for (const item of items) {
+    bucketSpecificAssignments.set(item.id, []);
+    bucketIndependantAssignments.set(item.id, []);
+  }
+
+  const itemSocketMetadata = _.mapValues(
+    _.keyBy(items, (item) => item.id),
+    (item) => getSpecialtySocketMetadatas(item)
+  );
+
+  const generalMods: PluggableInventoryItemDefinition[] = [];
+  const combatMods: PluggableInventoryItemDefinition[] = [];
+  const raidMods: PluggableInventoryItemDefinition[] = [];
+
+  for (const mod of mods) {
+    if (mod.plug.plugCategoryHash === armor2PlugCategoryHashesByName.general) {
+      generalMods.push(mod);
+    } else if (combatCompatiblePlugCategoryHashes.includes(mod.plug.plugCategoryHash)) {
+      combatMods.push(mod);
+    } else if (raidModPlugCategoryHashes.includes(mod.plug.plugCategoryHash)) {
+      raidMods.push(mod);
+    } else {
+      const itemForMod = items.find(
+        (item) => mod.plug.plugCategoryHash === bucketsToCategories[item.bucket.hash]
+      );
+      itemForMod && bucketSpecificAssignments.get(itemForMod.id)?.push(mod);
+    }
+  }
+
+  const itemEnergies = _.mapValues(
+    _.keyBy(items, (item) => item.id),
```

Comment:
Why not map the items array and *then* keyBy, instead of having to use mapValues?

Reply:
Because we would lose the item and thus the item id at that point. I am mapping to specialty socket metadata which loses the item information.

Reply:
Ah I see